compiler/rustc_type_ir/src/elaborate.rs

```diff
@@ -89,6 +90,55 @@ impl<I: Interner, O: Elaboratable<I>> Elaborator<I, O> {
                     return;
                 }
 
+                // N.B. this is a bit of a hack to get implied bounds for effects working..
```

Comment:
```
// HACK(effects):
```

Reply:
Also, yeah, why does this even need to be here? Why isn't adding that bound in `push_trait_bound` sufficient?

Reply:
Please also pull this out into a separate method.

Reply:
This is because `<T as Tr>::Fx: EffectsCompat<somebool>` will elaborate to `<T as SuperTr>::Fx: EffectsCompat<somebool>`, which itself can elaborate to `<T as SuperSuperTr>::Fx: EffectsCompat<somebool>`. It isn't performance effective for us to recursively elaborate all bounds as the use site to all bounds that it implies.

Reply:
Nothing about the comment..

> N.B. this is a bit of a hack to get implied bounds for effects working..

..implies that this has anything to do with performance. I would appreciate if this were recorded more carefully somewhere.

---

I guess I could ask the same question but backwards, why does this logic need to exist in `push_trait_bound`? Is there any way that this could be located all in elaboration?

Reply:
Added some of the reasoning into the `HACK` comment, cc https://github.com/rust-lang/rust/pull/129499#issuecomment-2365239147